server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionMultiListenerTests.java

```diff
@@ -121,45 +123,60 @@ private static void awaitQuietly(CountDownLatch latch) {
 
     public void testShouldExecuteWithCorrectContext() {
 
-        var context = new ThreadContext(Settings.EMPTY);
-        var listener = new AllocationActionMultiListener<Integer>(context);
-
-        context.putHeader("header", "root");
-        var r1 = new AtomicReference<String>();
-        var r2 = new AtomicReference<String>();
-        var l1 = listener.delay(
-            ActionListener.wrap(
-                response -> r1.set(context.getHeader("header")),
-                exception -> { throw new AssertionError("Should not fail in test"); }
-            )
-        );
-        var l2 = listener.delay(
-            ActionListener.wrap(
-                response -> r2.set(context.getHeader("header")),
-                exception -> { throw new AssertionError("Should not fail in test"); }
-            )
-        );
+        final var requestHeaderName = "header";
+        final var responseHeaderName = "responseHeader";
 
-        executeInRandomOrder(
-            context,
-            List.of(
-                new Tuple<>("clusterStateUpdate1", () -> l1.onResponse(1)),
-                new Tuple<>("clusterStateUpdate2", () -> l2.onResponse(2)),
-                new Tuple<>("reroute", () -> listener.reroute().onResponse(null))
-            )
-        );
+        final var expectedRequestHeader = randomAlphaOfLength(10);
+        final var expectedResponseHeader = randomAlphaOfLength(10);
 
-        assertThat(r1.get(), equalTo("root"));
-        assertThat(r2.get(), equalTo("root"));
-    }
+        var context = new ThreadContext(Settings.EMPTY);
+        var listener = new AllocationActionMultiListener<>(context);
+
+        context.putHeader(requestHeaderName, expectedRequestHeader);
+        context.addResponseHeader(responseHeaderName, expectedResponseHeader);
+
+        var isComplete = new AtomicBoolean();
+        try (var refs = new RefCountingRunnable(() -> assertTrue(isComplete.compareAndSet(false, true)))) {
+
+            List<Runnable> actions = new ArrayList<>();
+
+            for (int i = between(0, 5); i > 0; i--) {
+                var expectedVal = new Object();
+                var delayedListener = listener.delay(
+                    ActionListener.releaseAfter(ActionListener.wrap(Assert::fail).delegateFailure((l, val) -> {
```

Comment:
I am not sure I understand `ActionListener.wrap(Assert::fail).delegateFailure((l, val) -> {` part.
Why not have just:
`ActionListener.wrap(..` with assertions below?

Reply:
We don't need the two-arg `ActionListener.wrap` because we don't need to feed exceptions from `onResponse` into `onFailure`. Also it seems simpler to me to assert things straight away?

Reply:
We don't need the two-arg `ActionListener.wrap` because we don't need to feed exceptions from `onResponse` into `onFailure`. Also it seems simpler to me to assert things straight away?